theano/tensor/extra_ops.py

```diff
@@ -1006,3 +1007,69 @@ def to_one_hot(y, nb_class, dtype=None):
     ret = theano.tensor.set_subtensor(ret[theano.tensor.arange(y.shape[0]), y],
                                       1)
     return ret
+
+class Unique(theano.Op):
+    __props__ = ("return_index", "return_inverse", "return_counts")
+
+    def __init__(self, return_index=False, return_inverse=False, 
+				 return_counts=False):
+        self.return_index = return_index
+        self.return_inverse = return_inverse
+        self.return_counts = return_counts
+        if self.return_counts == True and np.__version__ < "1.9.0" :
+            raise RuntimeError(
+			    "Numpy version = " + np.__version__ +
+				". Option 'return_counts=True' works starting"
+                " from version 1.9.0.")
+
+    def make_node(self, x):
+        x = T.as_tensor_variable(x)
+        #x = x.flatten()
+        
+        outputs = []
+        # output0 = T.TensorType(broadcastable=[False], dtype=x.dtype)()
+        output0 = x.flatten().type()
+        outputs.append(output0)
+        typ = T.TensorType(broadcastable=[False], dtype='int64')
+        if self.return_index :
+            output1 = typ()
+            outputs.append(output1)
+        if self.return_inverse :
+            output2 = typ()#T.TensorType(broadcastable=[False], dtype=x.dtype)
+            outputs.append(output2)
+        if self.return_counts :
+            output3 = typ()#T.TensorType(broadcastable=[False], dtype=x.dtype)
+            outputs.append(output3)
+            
+        return theano.Apply(self, [x], outputs)
+
+    def perform(self, node, inputs, output_storage):
+        x = inputs[0]
+        z = output_storage
+        param = {}       
+        if self.return_index : 
+            param['return_index'] = True
+        if self.return_inverse :
+            param['return_inverse'] = True
+        if self.return_counts:
+            param['return_counts'] = True
+        outs = np.unique(x,**param)
+        if ((not self.return_inverse) and
+                (not self.return_index) and
+		        (not self.return_counts)):
+            z[0][0]=outs
+        else :
+            for i in range(len(outs)): 
+                z[i][0] = outs[i]
+                
+    def infer_shape(self, node, i0_shapes):
+        ret = node.fgraph.shape_feature.default_infer_shape(node, i0_shapes)
+        if self.return_inverse :
+            shape = (np.prod(i0_shapes[0]), )
```

Comment:
use basic.prod(), this is symbolic shape. np.prod() should work, but make it less clear we work on symbolic shape.